---
title: '"Pipeline of Differential Expression Analysis whit Glimma of DLBCL data'
author: "Epit√°cio Farias, Beatriz Stransky, Patrick Terrematte"
date: "29/01/2021"
output: 
  html_document: 
    default
  github_document: 
    df_print: paged
    html_preview: FALSE
    keep_html: TRUE
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_format = "all") }) 
---

Full pipeline of DEA with Glimma - Comparing gene expression by Neoplasm Cancer Status (ajcc) of DLBCL - Diffuse Large B-Cell Lymphoma

https://www.cbioportal.org/study/summary?id=dlbc_tcga_pan_can_atlas_2018


```{r, error=TRUE, message=FALSE, warning=FALSE, purl=FALSE, results='hide'}
## This chunk automatically generates a text .R version of this script when running within knitr.
input  = knitr::current_input()  # filename of input document
output = paste(tools::file_path_sans_ext(input), 'R', sep = '.')
try(knitr::purl(input,output,documentation=2,quiet=T), silent = T)
# Avoid duplicate label error of knitr::purl
options(knitr.duplicate.label = 'allow')
# Code to browse the markdown file with renderized images.
knitr::opts_chunk$set(
  fig.path = "figs/render-"
)
```



# Installing and Loading Libraries            


```{r message=FALSE, warning=FALSE, paged.print=FALSE}

packages_bioconductor = c("DESeq2", "Glimma", "limma", "genefilter", "org.Hs.eg.db")

#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from Bioconductor and loaded
package.check <- lapply(packages_bioconductor, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

packages_cran = c("DT", "tidyverse","dplyr", "tibble", "stringr", "data.table", "genefilter", "ggrepel")
  
#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from CRAN and loaded
package.check <- lapply(packages_cran, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

rm(packages_cran, packages_bioconductor, package.check)
```


## Downloading Dataframes

```{r}
# https://github.com/cBioPortal/datahub/tree/master/public/

url <- "https://media.githubusercontent.com/media/cBioPortal/datahub/master/public/dlbc_tcga_pan_can_atlas_2018/data_RNA_Seq_v2_expression_median.txt"
destfile <- "data/dlbc_exp.txt"
download.file(url, destfile)


url <- "https://media.githubusercontent.com/media/cBioPortal/datahub/master/public/dlbc_tcga_pan_can_atlas_2018/data_clinical_patient.txt"
destfile <- "data/data_clinical.txt"
download.file(url, destfile)


```

##Loading data's

```{r}

data_clinical <- read.delim("../../data/data_clinical.txt", comment.char="#", na.strings=c("NA", ""))

dlbc_exp <- read.delim("../../data/dlbc_exp.txt", comment.char="#", na.strings=c("NA", ""))

dlbc_exp <- dlbc_exp[!is.na(dlbc_exp$Hugo_Symbol),]
dlbc_exp <- dlbc_exp[!duplicated(dlbc_exp$Hugo_Symbol),]
rownames(dlbc_exp) <- as.character(dlbc_exp$Hugo_Symbol)
dlbc_exp$Entrez_Gene_Id <- NULL
dlbc_exp$Hugo_Symbol <- NULL

```


```{r}
library(org.Hs.eg.db)

genenames <- row.names(dlbc_exp)

cols <- c("SYMBOL", "ENSEMBL", "GENENAME")

dlbc_anno <- AnnotationDbi::select(org.Hs.eg.db, keys=genenames, columns=cols, keytype="SYMBOL")

colnames(dlbc_anno) <- c("symbol", "ensembl", "description")
```


```{r}

data_ajcc <- data_clinical %>% 
              dplyr::filter(PERSON_NEOPLASM_CANCER_STATUS %in% c("Tumor Free", "With Tumor")) %>% 
              droplevels()


data_ajcc$codes <- gsub("-", "\\.", data_ajcc$PATIENT_ID)

colnames(dlbc_exp) <- substr(colnames(dlbc_exp),1,12)

data_ajcc$PERSON_NEOPLASM_CANCER_STATUS <- ifelse(data_ajcc$PERSON_NEOPLASM_CANCER_STATUS =="With Tumor", "TP","NT")
  
exp_ajcc <- dlbc_exp[,data_ajcc$codes]


exp_ajcc <- round(exp_ajcc, digits = 0)
```



# Running DESeq2 Differential Expression


```{r}

# create the DESeqDataSet object
ddsObj <- DESeqDataSetFromMatrix(countData = exp_ajcc,
                              colData = data_ajcc,
                              design = ~ PERSON_NEOPLASM_CANCER_STATUS )
```


# Interactive StripChart with Glimma
    
    
```{r}
ddsObj <- DESeq(ddsObj)
res.DESeq2 <- results(ddsObj)
res.shr <- DESeq2::lfcShrink(ddsObj, coef=2, res=res.DESeq2)

```


```{r}
summary(res.shr)

```

```{r}
shrink.deseq <- as.data.frame(res.shr) %>%
    rownames_to_column("symbol") %>% 
    left_join(dlbc_anno, by="symbol") %>% 
    rename(logFC=log2FoldChange, FDR=padj)
```


```{r}
df.deseq <- shrink.deseq[abs(shrink.deseq$logFC) > 0.5 & shrink.deseq$FDR < 0.1,]

df.deseq <- df.deseq[!is.na(df.deseq$logFC),]
dim(df.deseq)
```


```{r}
genes.DEA.FT.vs.WT.lst <- unique(df.deseq$symbol)
genes.DEA.FT.vs.WT.lst <- genes.DEA.M0.vs.M1.lst[!grepl('^LOC', genes.DEA.FT.vs.WT.lst)]
write(genes.DEA.FT.vs.WT.lst,  file = "genes.DEA.FT.vs.WT.lst")
```

```{r}

save.image("DLBCL_pipeline_glimma_DEA_count.RData")

``